### ISSUE TITLE
cp oom

### ISSUE THREAD

**Issue by sunzx8** [THUDM/slime#1854]
# [RFC] Kernel-aware Context Parallel (KCP) for FLA-backed linear attention models

## Summary

I'd like to propose completing the `--allgather-cp` code path by pushing context
parallelism **into the FLA kernel** on attention layers that support it
(currently `qwen3_5` and `qwen3_next`). Today the flag only affects data
partitioning (`data.py`) and loss redistribution (`loss.py`); the attention
layer itself still runs the original zigzag all-gather + duplicated-compute
path, which is the dominant source of activation blow-up at long context.

We've prototyped this end-to-end and ran a **4h26min, 28-rollout production
training on Qwen3.5-122B-A10B at 180k context on 304 GPUs (38 nodes)**. Results
below. Before we upstream the patch I wanted to check with maintainers that
(a) this direction is welcome, (b) there isn't already a WIP branch doing the
same thing, and (c) the proposed API is aligned with how you want this to
evolve.

If the direction is green-lit, we'll follow up with a proper PR (with tests
and benchmarks).

---

### 1.1 Current state of `--allgather-cp` in `main`

`args.allgather_cp` today is plumbed through:

- `slime/backends/megatron_utils/data.py` — DSA-style partition: global
  concat → pad to `cp_size * pad_size` → `tokens.chunk(cp_size)[cp_rank]`.
  Each rank ends up with a **contiguous** slice of the global packed stream.
  `cu_seqlens` is kept **global** (its last entry equals the global padded
  length, not the per-rank length).
- `slime/backends/megatron_utils/loss.py` — `_allgather_cp_redistribute` + the
  `args.allgather_cp` branches already understand contiguous per-rank logits
  layout.
- `slime/backends/megatron_utils/model.py` — `args.allgather_cp` is forwarded.

### 1.2 What's missing in `main`

`slime_plugins/models/hf_attention.py` has no `allgather_cp` branch. The only
CP code path is:

```python
# slime_plugins/models/hf_attention.py (current main)
if mpu.get_context_parallel_world_size() > 1:
    cp_size = mpu.get_context_parallel_world_size()
    hidden_states_list = _AllGatherForDuplicatedComputation.apply(
        hidden_states, mpu.get_context_parallel_group(),
    )
    local_cu_seqlens = cu_seqlens // cp_size      # <-- assumes zigzag partition
    for i in range(len(cu_seqlens) - 1):
        seqlen = cu_seqlens[i + 1] - cu_seqlens[i]
        chunk_size = seqlen // 2 // cp_size       # <-- assumes zigzag partition
        ...
    hidden_states = torch.cat(whole_hidden_states_list, dim=0)
```

The highlighted lines assume the zigzag partition (i.e. `allgather_cp=False`).
When `data.py` produces the **contiguous** `allgather_cp=True` layout with a
**global** `cu_seqlens`, the attention path is still interpreting its inputs
under the zigzag assumption. This creates a semantic/layout gap between the
data path and the attention path; depending on the packed layout, that gap may
lead to incorrect slicing or shape issues.

In other words: **`--allgather-cp` has already been propagated through the
`data.py` / `loss.py` path on `main`, but not yet through the attention
layer.** This RFC proposes to close that gap, and while doing so, to do it in
a way that also eliminates the outer activation blow-up on FLA-backed models.

### 1.3 Why "kernel-aware" instead of "outer all-gather + duplicated compute"

For attention kernels that can internally do their own CP state exchange
(`ShortConvolution` + `chunk_gated_delta_rule` and friends in FLA), the outer
all-gather is pure overhead:

- Peak activation grows to the full global sequence length on every CP rank
  (that's the OOM we hit at 180k on Qwen3.5-122B even with PP=8 CP=2 EP=8 and
  `recompute-granularity full`).
- Every rank re-computes the same attention — `O(cp_size)` wasted compute.
- After attention we immediately slice it back down — the gather was only
  needed because the kernel didn't know how to communicate across CP ranks.

`fla.ops.cp.build_cp_context` exists precisely to let FLA kernels handle their
own halo / state exchange when given a `CONTEXT_PARALLEL_GROUP` and the
global `cu_seqlens`. KCP simply wires this in.

---

### 2.1 Minimum-invasive extension points on `HuggingfaceAttention`

Two new hooks, both defaulting to "not supported" so existing attention
subclasses (GLM, gpt_oss, …) keep the legacy path with zero behavior change:

```python
class HuggingfaceAttention(MegatronModule, ABC):

    def supports_fla_kcp(self) -> bool:
        """Whether this attention module supports FLA Kernel-aware CP."""
        return False

    def get_cp_conv_kernel_size(self) -> int | None:
        """conv1d kernel size used by this module, for building FLACPContext."""
        return None
```

Subclasses whose `hf_forward` internally calls FLA kernels that understand
`cp_context` override these. For example:

```python
class Qwen3_5GatedDeltaNet(HuggingfaceAttention):
    def supports_fla_kcp(self) -> bool:
        return True
    def get_cp_conv_kernel_size(self) -> int | None:
        return self.linear_attn.conv_kernel_size
```

### 2.2 Routing

A small router decides per-forward whether to take the KCP fast path:

```python
def _should_use_fla_kcp(self) -> bool:
    if mpu.get_context_parallel_world_size() <= 1:
        return False
    if not self.supports_fla_kcp():
        return False
    if not getattr(self.args, "allgather_cp", False):
        return False
    return True
```

All three conditions must hold. If any is false, the module falls back to the
existing legacy CP path — **no change for modules that don't opt in, no
change for `allgather_cp=False` runs**.

### 2.3 Forward path

When `_should_use_fla_kcp()` is true:

```python
from fla.ops.cp import build_cp_context

cp_context = build_cp_context(
    cu_seqlens=cu_seqlens,                            # global, from data.py
    group=mpu.get_context_parallel_group(),
    conv1d_kernel_size=self.get_cp_conv_kernel_size(),
)
# hidden_states is already this rank's contiguous slice
# (global[r*T/C : (r+1)*T/C)), matching what data.py produced.
output = self.hf_forward(hidden_states, packed_seq_params, cp_context=cp_context)
```

When KCP is active, `hf_forward` must **not** additionally pass `cu_seqlens`
to the underlying FLA kernels — `cp_context` already carries the rank-local
cu_seqlens and halo metadata. Double-specifying would confuse the kernels.
(We handle this by passing `cu_seqlens=None` in the subclass when
`cp_context is not None`; see §2.4.)

### 2.4 Subclass changes (Qwen3.5 / Qwen3-Next)

Signature of `hf_forward` gains an optional `cp_context=None`. Inside the
FLA kernel calls:

```python
kernel_cu_seqlens = None if cp_context is not None else cu_seqlens
kernel_cp_kwargs  = {"cp_context": cp_context} if cp_context is not None else {}
...
fla_kernel(
    ...,
    cu_seqlens=kernel_cu_seqlens,
    **kernel_cp_kwargs,
)
```

Other HF attention subclasses (GLM, gpt_oss, …) require **no change** — they
keep `supports_fla_kcp() == False`.

### 2.5 `data.py` — optional sanity assert

The `allgather_cp` branch in `data.py` is already doing the right thing. We
only add a small post-condition assert so that if anyone ever breaks the
padding invariant, KCP will fail loudly at data time instead of silently
mis-routing inside the kernel:

```python
assert cu_seqlens[-1].item() == tokens.size(0) * cp_size, (
    f"allgather_cp sanity: expected global cu_seqlens[-1]="
    f"{tokens.size(0) * cp_size}, got {cu_seqlens[-1].item()}"
)
```

### 2.6 Fallback / escape hatches

- `allgather_cp=False` → `_should_use_fla_kcp` returns False → unchanged
  zigzag legacy path.
- `supports_fla_kcp()` returns False (non-FLA models) → unchanged legacy
  path.
- `cp_size == 1` → unchanged single-rank path.

No new CLI flag is introduced. `--allgather-cp` becomes the single switch
that, for FLA-backed models, now fully propagates from data → attention →
loss as one coherent DSA path.

---

### 3.1 Closing the `allgather_cp` correctness gap

Even independent of performance: on `main` today, running
`--allgather-cp` with `cp_size > 1` on an FLA-backed linear-attention model
still enters an attention code path (`local_cu_seqlens = cu_seqlens // cp_size`)
whose invariants are written for zigzag partitioning. That semantic mismatch is
what this RFC resolves.

### 3.2 Activation memory at long context

For FLA-backed models the outer gather is eliminated. The two
FLA `ref_log_probs` / `log_probs` forward passes at 180k context no longer
need to materialize the global sequence on each CP rank — they go from ~55s
per pass to ~15–17s per pass (measured, see §4). `actor_train` stops OOM'ing
at 180k under the same parallelism that used to OOM with legacy CP.

### 3.3 Duplicated-compute factor of `cp_size`

Linear attention on each rank now processes only its own `T/cp_size` slice
(plus halo tokens), not the full `T` — saving roughly `O(cp_size)` FLOPs at
the attention layer.

---

## 4. Validation

Measured on our setup:

- Model: **Qwen3.5-122B-A10B** (48 layers, linear attention via
  `chunk_gated_delta_rule`, `num_query_groups=2`)
- Hardware: **38 nodes × 8 × A800-80GB = 304 GPUs**
- Parallelism: `TP=2 PP=8 CP=2 EP=8 ETP=1` → DP=8
- Context: **real 180k per sample** (`total_lengths ≈ 180086`)
- Flags: `--allgather-cp`, `--micro-batch-size 1`,
  `--recompute-granularity full --recompute-num-layers 1`,
  `--optimizer-cpu-offload`, `--max-tokens-per-gpu 100000`

### 4.1 Functional pass

Single-rollout smoke (v28):

| Stage | Elapsed |
|---|---|
| `update_weights` | 20.5 s |
| Rollout (sglang generate 180k) | ~80 s |
| `ref_log_probs` (KCP fwd) | **17.5 s** |
| `log_probs` (KCP fwd) | **15.4 s** |
| `actor_train` (KCP fwd + bwd) | **703.3 s** |
| `save_model` | 96.7 s |
| `update_weights` | 14.6 s |
| Job status | SUCCEEDED |

### 4.2 Stability (28 rollouts, 4h26min)

| Metric | Value |
|---|---|
| Rollouts completed | **28 / 28** |
| actor_train steady-state mean (rollouts 1-27) | **382.0 s** |
| actor_train steady-state stddev | **4.6 s (1.2 %)** |
| First rollout (incl. Triton JIT + autotune) | 413.9 s |
| Last rollout (r27) | 385.4 s → **no drift** |
| `save_model` (6 checkpoints, every 5 rollouts) | 94.6 – 97.2 s |
| OOM / NCCL watchdog / Triton errors / tracebacks | **0** |

The log_probs / ref_log_probs stay at ~15 s for all 28 rollouts, confirming
the KCP forward path is cache-stable across the whole run.

### 4.3 Counter-factual

On the same hardware and same parallelism:

- On our setup, the legacy CP path (without KCP) OOM'd during `actor_train`
  at 180k with `71.66 GiB allocated by PyTorch` at the linear-attn layer.
  This was our baseline failure mode across ~20 runs before KCP.
- On the same setup, increasing CP from 2 to 4 did **not** help, because the
  outer all-gather restored the full-length activation regardless of `cp_size`.
- Raising TP was not available for this model: Qwen3.5-122B has
  `num_query_groups=2`, which Megatron requires to be divisible by TP, so
  TP ≤ 2.

So for this workload, on this hardware, we were not able to reach stable 180k
training without a kernel-aware CP path.

---

## 5. Questions for maintainers

1. **Is there already a WIP branch in this direction?** We saw that
   `allgather_cp` has been plumbed through `data.py` / `loss.py` /
   `model.py` but not through `hf_attention.py`. If a KCP (or equivalent)
   patch is already in flight, we'd rather align with it than duplicate
   effort.
2. **API shape for the capability hook.** We modeled `supports_fla_kcp()`
   + `get_cp_conv_kernel_size()` as instance methods on
   `HuggingfaceAttention`. If you prefer a different shape (class-level
   attribute, registry, arg plumbed through config, …) we're happy to
   reshape before PR.
3. **Scope for the first PR.** We see two sensible sizes:
   - **Narrow**: only the `hf_attention.py` routing + hooks + the two
     already-supported subclasses (`qwen3_5`, `qwen3_next`) + sanity assert
     + tests. No behavior change for any other model.
   - **Wider**: also include the `data.py` assert and any small cleanup in
     `loss.py`'s `allgather_cp` branches we noticed along the way.

   We'd default to Narrow unless you prefer Wider.
4. **Testing expectations.** We have a 2-GPU numerical correctness smoke
   test (KCP vs legacy `out` / `dhidden` with bf16 tolerance `< 5e-2`) and a
   CP=2 routing test. Are CP=4 / CP=8 tests required in CI? Anything else
   you'd want to see before PR?
5. **Is there an existing `rollout/ref_log_probs` speedup benchmark in the
   repo** we should compare against, so the PR includes a familiar number
   format rather than our internal one?

---

## 6. Proposed next step

If the direction is OK, we'll follow up with a PR containing:

- `slime_plugins/models/hf_attention.py` — KCP router + two hooks + KCP
  fast path; legacy path untouched for modules that don't opt in.
- `slime_plugins/models/qwen3_5.py`, `slime_plugins/models/qwen3_next.py`
  — opt-in via the two hooks; `hf_forward` grows an optional
  `cp_context=None` argument.
- `slime/backends/megatron_utils/data.py` — one sanity assert on the
  `allgather_cp` branch.
- `tests/test_qwen3_linear_attention_cu_seqlens.py` — expanded to cover
  KCP routing.
- `tests/test_qwen3_fla_kcp_smoke.py` — new 2-GPU numerical smoke.

Total diff is on the order of ~250 added / ~40 changed lines, plus tests.

Thanks for reading — looking forward to your feedback on direction and shape
before we send the PR.

---
